library/Network/CloudSeeder/Main.hs

```diff
@@ -98,53 +125,131 @@ runAppM (AppM x) = do
+getTags config stackToDeploy env appName = baseTags <> globalTags <> localTags
+  where 
+    baseTags :: S.Set (T.Text, T.Text)
+    baseTags = [("cj:environment", env), ("cj:application", appName)]
+    globalTags = config ^. tagSet
+    localTags = stackToDeploy ^. tagSet
+
+getPassedParameters :: ( MonadArguments e f, HasParameterSources s1 (S.Set (T.Text, ParameterSource)), HasParameterSources s (S.Set (T.Text, ParameterSource))) 
+                    => s1 -> s -> f (S.Set (T.Text, T.Text))
+getPassedParameters config stackToDeploy = do
+  let globalParamSources = config ^. parameterSources
+      localParamSources = stackToDeploy ^. parameterSources
+      isFlag (_, Flag) = True
+      isFlag (_, _) = False
+      paramFlags = S.fromList $ (globalParamSources <> localParamSources)^..folded.filtered isFlag
+      --TODO: use a more lensy approach, a la something like replacing "isFlag" w/ "(_2.is _Flag)"
+  S.fromList . M.toList <$> getOptions paramFlags
 
-  let globalParams = (config ^. parameters) ++ [("Env", env)]
+collectParameters :: (HasParameters s1 (S.Set (T.Text, T.Text)), HasParameters s (S.Set (T.Text, T.Text))) 
+                  => s1 -> s -> S.Set (T.Text, T.Text) -> T.Text -> S.Set (T.Text, T.Text) -> S.Set (T.Text, T.Text) -> S.Set (T.Text, T.Text)
+collectParameters config stackToDeploy envVars env passedParams outputs =
+  let globalParams = config ^. parameters
       localParams = stackToDeploy ^. parameters
-      allParams = globalParams ++ localParams
-      isRequired = (`elem` requiredParams)
-      stackParams = filter (isRequired . fst) (allParams ++ outputs ++ envVars)
+  in globalParams <> localParams <> outputs <> envVars <> [("Env", env)] <> passedParams
 
-  csId <- computeChangeset (mkStackName nameToDeploy) templateBody stackParams stackTags
-  runChangeSet csId
+getParameters :: (MonadArguments CliError f, HasEnvironmentVariables s [T.Text], HasEnvironmentVariables s1 [T.Text], MonadEnvironment f, MonadCloud f, Traversable t, HasParameterSources s1 (S.Set (T.Text, ParameterSource)), HasParameterSources s (S.Set (T.Text, ParameterSource)), HasParameters s1 (S.Set (T.Text, T.Text)), HasParameters s (S.Set (T.Text, T.Text))) 
+              => s1 -> s -> t T.Text -> T.Text -> T.Text -> f (S.Set (T.Text, T.Text))
+getParameters config stackToDeploy dependencies env appName = do 
+  envVars <- getEnvVars config stackToDeploy
+  outputs <- getOutputs dependencies env appName
+  passedParams <- getPassedParameters config stackToDeploy
+  return $ collectParameters config stackToDeploy envVars env passedParams outputs
+
+validateTags :: (MonadError e m, AsCliError e) => S.Set (T.Text, T.Text) -> m (M.Map T.Text T.Text)
+validateTags ts = assertUnique _CliDuplicateTagValues ts
+
+validateParameters :: (AsCliError e, HasParameterSpecs s (f ParameterSpec), Foldable f, MonadError e m) 
```

Comment:
Holy genericism, batman! I’m not sure making this generic over an arbitrary foldable is necessary… do we actually use that power? Or do we always make `f` the same thing where we call this function?